FancyBatching.md

```diff
@@ -0,0 +1,505 @@
+
+    - If `AutoFlush` is `true`
+        - Call `TrySubmit` if the batch is full.  If `TrySubmit` returns `false`,
+        we'll just return because someone else is submitting.  If `TrySubmit` is
+        `true` and there's pending actions remaining we'll start the timer.
+
+- Flushing
+    - Call Validate
+    - The `Flush`/`Async` methods will call `TrySubmit(flush: true)`.  If that returns `false`
+    because another thread is already submitting, it will set a `_pendingFlush`
+    flag to `true`.  After the submission is finished, we'd normally check
+    whether another batch was ready to go.  If `_pendingFlush` is true, we'll
+    submit any remaining actions even if they're not a full batch.
+
+- Disposing
+    - Call Validate
+    - The `Dispose`/`DisposeAsync` methods will call `TrySubmit(block: true)` that
+    block on submission.  We'll also set a `_disposed` flag on the publisher.
+    Make sure calling `Dispose` from every sender thread works safely.
+
+- `Task<bool> TrySubmitAsync(bool flush, bool block)`
+    - First it will make sure it's the only thread currently submitting.  It will
+    use `Monitor.TryEnter` when `block == false` and `Monitor.Enter` otherwise.
+        - If we can't acquire the monitor, we'll return `false`.
+        - If we do acquire the monitor, we'll check if we've been `_disposed` and
+        return `false` (in case multiple `Dispose` calls were all waiting on the
+        monitor).
+
+    - Disable any existing timer
+
+    - If `flush || (_pendingSize + _retrySize >= BatchSize)`
+        - Fill the batch
+        - `SubmitBatchAsync(batch, retryAttempt: 0)`
+        - Set `_pendingFlush` to `false`
+
+    - Call `Monitor.Exit` and return `true`
+
+- `Task SubmitAsync(Batch<T> batch, int retryAttempt)`
+    - Raise the ActionSent event for every item in batch
+    - Send the filled batch as a request using our `SearchClient`.
+
+    - If we get a 413 Payload Too Large
+        - Keep adding items to fill a new batch so long as we're less than half
+        the size of the payload.  Put everything else on the `_retryActions`
+        queue but don't increment their `RetryCount`.  Call `SubmitAsync()`
+        recursively without changing the `retryAttempt`.
+
+    - If we get a 200 or 207
+        - Associate errors with their actions from the batch
+        - For each error
+            - If the `RetryCount` is less than the max, increment the count and
+            add it to the `_retryActions` queue
+            - Else raise the `ActionFailed` event and log the failure
+        - For each success
+            - Raise the `ActionCompleted` event
+
+    - If `_pendingFlush || (_pendingSize + _retrySize >= BatchSize)`
+        - Fill the batch
+        - If any of the errors were 503s
+            - Add an exponential retry delay with hitter based on the number of
+            attempts (follow your `RetryPolicy`'s implementation)
+            - call `SubmitBatch(batch, retryAttempt + 1)`
```

Comment:
Is this retry attempt `Math.max` of all documents being retried? Say we get a batch with a document on its 3rd attempt and all others on their first attempt we'll submit with `3+1`?

Reply:
Not in my current thinking.  If you had a flush resulting in a couple of retries pending but not enough to trigger another auto-flush, the fact those were retried twice already doesn't say anything about when the next batch is sent.

Reply:
We should consider changing this to keep track of when the last response was sent and when the retry backoff would finish.  If we submit the next request before the backoff would have elapsed, we can wait that amount of time before sending the next request.